rcl_action/src/rcl_action/goal_handle.c

```diff
@@ -0,0 +1,162 @@
+  RCL_CHECK_ALLOCATOR_WITH_MSG(&allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Ensure the goal handle is zero initialized
+  if (goal_handle->impl) {
+    RCL_SET_ERROR_MSG("goal_handle already initialized, or memory was unintialized");
+    return RCL_RET_ALREADY_INIT;
+  }
+  // Allocate space for the goal handle impl
+  goal_handle->impl = (rcl_action_goal_handle_impl_t *)allocator.allocate(
+    sizeof(rcl_action_goal_handle_impl_t), allocator.state);
+  if (!goal_handle->impl) {
+    RCL_SET_ERROR_MSG("goal_handle memory allocation failed");
+    return RCL_RET_BAD_ALLOC;
+  }
+  // Allocate space for the goal info pointer
+  goal_handle->impl->info = (rcl_action_goal_info_t *)allocator.allocate(
+    sizeof(rcl_action_goal_info_t), allocator.state);
+  if (!goal_handle->impl->info) {
+    RCL_SET_ERROR_MSG("goal_handle info memory allocation failed");
+    return RCL_RET_BAD_ALLOC;
+  }
+  // Copy goal info and allocator into goal handle
+  *goal_handle->impl->info = *goal_info;
+  goal_handle->impl->allocator = allocator;
+  // Initialize state to ACCEPTED
+  goal_handle->impl->state = GOAL_STATE_ACCEPTED;
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_goal_handle_fini(rcl_action_goal_handle_t * goal_handle)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handle, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_action_goal_handle_is_valid(goal_handle)) {
+    return RCL_RET_ACTION_GOAL_HANDLE_INVALID;
+  }
+  goal_handle->impl->allocator.deallocate(
+    goal_handle->impl->info, goal_handle->impl->allocator.state);
+  goal_handle->impl->allocator.deallocate(goal_handle->impl, goal_handle->impl->allocator.state);
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_update_goal_state(
+  rcl_action_goal_handle_t * goal_handle,
+  const rcl_action_goal_event_t goal_event)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handle, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_action_goal_handle_is_valid(goal_handle)) {
+    return RCL_RET_ACTION_GOAL_HANDLE_INVALID;
+  }
+  rcl_action_goal_state_t new_state = rcl_action_transition_goal_state(
+    goal_handle->impl->state, goal_event);
+  if (GOAL_STATE_UNKNOWN == new_state) {
+    return RCL_RET_ACTION_GOAL_EVENT_INVALID;
+  }
+  goal_handle->impl->state = new_state;
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_goal_handle_get_info(
+  const rcl_action_goal_handle_t * goal_handle,
+  rcl_action_goal_info_t * goal_info)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handle, RCL_RET_INVALID_ARGUMENT);
+  if (!rcl_action_goal_handle_is_valid(goal_handle)) {
+    return RCL_RET_ACTION_GOAL_HANDLE_INVALID;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_info, RCL_RET_INVALID_ARGUMENT);
+  *goal_info = *goal_handle->impl->info;
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_goal_handle_get_status(
+  const rcl_action_goal_handle_t * goal_handle,
+  rcl_action_goal_state_t * status)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handle, RCL_RET_INVALID_ARGUMENT);
```

Comment:
nit, `rcl_action_goal_handle_is_valid()` also checks if `goal_handle` is `NULL`. This check could be removed.

Reply:
Then it's a good opportunity to make it consistent :)

Reply:
Removed the the check (and other similar redundant checks) in 8c65da8
Rather than distinguishing a null pointer vs uninitialized object by return code, we can use the error message instead.

Reply:
Follow-up for `rcl` #321